ruby/sdk-phase-2.md

```diff
@@ -0,0 +1,605 @@
+# Ruby - Phase 2
+
+This proposal is for implementing workflows in the Ruby SDK.
+
+See [phase 1](sdk-phase-1.md) for details on phase 1 which includes client and activities (granted some things may have
+changed since first authored).
+
+Notes for this proposal (like phase 1):
+
+* Everything is subject to change both during proposal time and afterwards during implementation.
+* For each place a decision is made, a "💭 Why X? ..." aside is present explaining why.
+* For each place an open question exists, a "❓ X?" exists to ask it.
+
+Here is a high-level overview of potentially controversial decisions:
+
+* Base class for activities is now `Temporalio::Activity::Definition` to be consistent with the base class
+* Class methods are dynamically made available for signals, queries, and updates to use from clients.
+  * So `MyWorkflow.my_signal` can be used in a client, and it just returns a signal definition, unlike the instance
+    form.
+* By default `TracePoint` will be enabled for workflow threads that will make sure users are not calling illegal calls,
+  but there is a way to disable for blocks of code.
+* `Ractor`s will be used for state isolation by default assuming decent performance but easy to disable and mostly
+  hidden from users.
+
+These points are explained below.
+
+## Workflow Definition/Usage
+
+Examples below are taken from
+[this Python sample](https://github.com/temporalio/samples-python/tree/main/message_passing/introduction), so not too
+many comments about general non-Ruby behavior are inline.
+
+Example definition:
+
+```ruby
+class CallGreetingServiceActivity < Temporalio::Activity::Definition
+  GREETINGS = {
+    arabic: 'مرحبا بالعالم',
+    chinese: '你好，世界',
+    english: 'Hello, world',
+    french: 'Bonjour, monde',
+    hindi: 'नमस्ते दुनिया',
+    portuguese: 'Olá mundo',
+    spanish: 'Hola mundo'
+  }
+
+  def execute(language)
+    # Simulate a network call
+    sleep(0.2)
+    GREETINGS[language.to_sym]
+  end
+end
+
+class GreetingWorkflow < Temporalio::Workflow::Definition
+  workflow_query_attr :language
+
+  def initialize(name)
```

Comment:
I don't think it should be the default and we chose not to make it the default in .NET even though `[WorkflowInit]` was present since its inception. There is a `workflow_init` below to facilitate opting in just like every other SDK that allows it, not make Ruby some special thing here. Would also like not to eagerly validate that the constructor arguments match the execute arguments if not opted in.

Reply:
TBC, what I'm proposing is just to require initialize to take start workflow args always.  That feels like the most normal and straightforward approach in any object-oriented setting, familiar to everyone.  Your approach will dramatically reduce the number of folks who take advantage.
I understand that from your perspective, the workflow init approach is "normal."  But stepping out of your Chad shoes, what we've done in other SDKs is pretty esoteric.  To most users, using the constructor is normal unless they happen to have used temporal in other languages, and even they will be plenty familiar with constructors.